data/transactions/logic/frames_test.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"testing"
+
+	"github.com/algorand/go-algorand/test/partitiontest"
+)
+
+const frameNonsense = `
+ double:
+  proto 1 1
+  pushn 1						// one return value
+  frame_dig -1
+  int 2
+  *
+  frame_bury 0
+  retsub
+  pushint 2
+  popn 1
+`
+
+const frameCompiled = "f00101f301f1ff240bf200898102f401"
+
+func TestPushPopN(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// These two are dumbs uses of pushn, and should perhaps be banned
+	testAccepts(t, "pushn 0; int 1", fpVersion)
+	testAccepts(t, "pushn 1; !", fpVersion)
+
+	// These two are equally dumbs uses of popn, and should perhaps be banned
+	testAccepts(t, "int 1; popn 0", fpVersion)
+	testAccepts(t, "int 1; dup; popn 1;", fpVersion)
+
+	testAccepts(t, "pushn 2; pop; !", fpVersion)
+	testAccepts(t, "pushn 3; !; assert; !; assert; !", fpVersion)
+	testPanics(t, "pushn 2", fpVersion)
+
+	testAccepts(t, "int 1; int 1; int 1; popn 2", fpVersion)
+	testAccepts(t, "int 1; int 0; popn 1", fpVersion)
+	testPanics(t, "int 1; int 0; popn 2", fpVersion)
+	testProg(t, "int 1; int 0; popn 3", LogicVersion, Expect{1, "popn 3 expects 3..."})
+	testPanics(t, notrack("int 1; int 0; popn 3"), fpVersion)
+
+	testAccepts(t, `pushn 250; pushn 250; pushn 250; pushn 250;
+                    popn 250;  popn 250;  popn 250;  popn 249; !`,
+		fpVersion)
+	testPanics(t, `pushn 250; pushn 250; pushn 250; pushn 251
+                   popn 250;  popn 250;  popn 250;  popn 250; !`,
+		fpVersion)
+}
+
+func TestPushPopNTyping(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	testProg(t, "pushn 2; +", LogicVersion)
+	testProg(t, "pushn 2; concat", LogicVersion, Expect{1, "...wanted type []byte..."})
+
+	testProg(t, "popn 1", LogicVersion, Expect{1, "...expects 1 stack argument..."})
+}
+
+func TestSimpleFrame(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	testAccepts(t, `
+        int 3
+        int 4
+        callsub hyp
+        int 5
+        ==
+        return
+      hyp:
+        proto 2 1
```

Comment:
They can not be negative in TEAL source, because they are "normal" immediates, not the new special kind that can be signed (only for `frame_dig` and `frame_bury`.

And then, once compiled to a `byte`, they are always considered positive.